elm-indent-simple.el

```
@@ -0,0 +1,171 @@
+  (save-excursion
+    (forward-line -1)
+    (back-to-indentation)
+    (looking-at-p (regexp-opt tokens))))
+
+(defun elm-indent-simple-compute-indentation ()
+  "Return a column to indent to.
+
+The numbers we get are the positions we can determine from the
+given context. When we cannot find a context to indent to, we
+default to the indentation level of previous line."
+  (let* ((indent-level-previous-line
+          (save-excursion
+            (forward-line -1)
+            (current-indentation)))
+         (positive-offset (+ indent-level-previous-line elm-indent-simple-offset)))
+    (save-excursion
+      (back-to-indentation)
+      ;; Now we are positioned at start of indentation.
+      ;; Logic below assumes this is true.
+      (cond
+       ((elm-indent-simple--previous-line-ends-with '("=" "<-" "[" "{" "of" "if" "else" ":" "->" "exposing")) positive-offset)
+       ((looking-at-p (regexp-opt '("{-" "-}"))) 0)
+       ((and (= indent-level-previous-line 0) (looking-at-p "=")) positive-offset)
+       ((elm-indent-simple--previous-line-starts-with '("type" "let")) positive-offset)
+       ((looking-at-p ")") (elm-indent-simple--find-indentation-of-list))
+       ((looking-at-p "}") (elm-indent-simple--find-indentation-of-list))
+       ((looking-at-p "]") (elm-indent-simple--find-indentation-of-list))
+       ((looking-at-p ",") (elm-indent-simple--find-indentation-of-list))
+       ((looking-at-p "else") (elm-indent-simple--find-indentation-of-tokens '("if")))
+       ((looking-at-p "then") (elm-indent-simple--find-indentation-of-tokens '("if")))
+       (t (elm-indent-simple-lastly))))))
+
+(defun elm-indent-simple-level-2-previous-lines ()
+  "Returns indent level of the two previous lines."
+  (save-excursion
+    ;; Since we save-excursion above, we can go back two lines one by one, and
+    ;; return the lines separately.
+    (cl-values
+     (progn
+       (forward-line -1)
+       (current-indentation))
+     (progn
+       (forward-line -1)
```

Note that sometimes there may not *be* one or even two previous lines, in which case `(forward-line -1)` would fail silently and you'd just get the indentation of the currently-reached line.